File: solutions/Python/construct_quad_tree_recursive.py
from typing import List

class Node:
    def __init__(self, val, isLeaf, topLeft=None, topRight=None, bottomLeft=None, bottomRight=None):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight

def construct(grid: List[List[int]]) -> 'Node':
    def dfs(top, left, size):
        all_same = True
        val = grid[top][left]
        
        for i in range(top, top + size):
            for j in range(left, left + size):
                if grid[i][j] != val:
                    all_same = False
                    break
            if not all_same:
                break
        
        if all_same:
            return Node(val == 1, True)
        
        half = size // 2
        topLeft = dfs(top, left, half)
        topRight = dfs(top, left + half, half)
        bottomLeft = dfs(top + half, left, half)
        bottomRight = dfs(top + half, left + half, half)
        
        return Node(1, False, topLeft, topRight, bottomLeft, bottomRight)
    
    return dfs(0, 0, len(grid))

grid = [[1,1],[1,0]]
root = construct(grid)
print(root.val, root.isLeaf)

File: solutions/Golang/construct_quad_tree_recursive.go
package main

import "fmt"

type Node struct {
	Val         bool
	IsLeaf      bool
	TopLeft     *Node
	TopRight    *Node
	BottomLeft  *Node
	BottomRight *Node
}

func construct(grid [][]int) *Node {
	return dfs(grid, 0, 0, len(grid))
}

func dfs(grid [][]int, top, left, size int) *Node {
	allSame := true
	val := grid[top][left]

	for i := top; i < top+size && allSame; i++ {
		for j := left; j < left+size && allSame; j++ {
			if grid[i][j] != val {
				allSame = false
			}
		}
	}

	if allSame {
		return &Node{Val: val == 1, IsLeaf: true}
	}

	half := size / 2
	node := &Node{Val: true, IsLeaf: false}
	node.TopLeft = dfs(grid, top, left, half)
	node.TopRight = dfs(grid, top, left+half, half)
	node.BottomLeft = dfs(grid, top+half, left, half)
	node.BottomRight = dfs(grid, top+half, left+half, half)
	return node
}

func main() {
	grid := [][]int{{1, 1}, {1, 0}}
	root := construct(grid)
	fmt.Println(root.Val, root.IsLeaf)
}
